File: 0143-reorder-list/0143-reorder-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
  const middle = slice(head);
  const second = reverse(middle);
  
  merge(head, second);
};

const slice = (node) => {
  let [slow, fast] = [node, node];
  
  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;
  }
  
  let next = slow.next;
  slow.next = null;
  
  return next;
}

const reverse = (node) => {
  let prev = null;
  
  while (node) [node.next, prev, node] = [prev, node, node.next];
  
  return prev;
}

const merge = (a, b) => {
  let [aNext, bNext] = [null, null];
  
  while (b) {
    aNext = a.next;
    bNext = b.next;
    
    a.next = b;
    b.next = aNext;
    
    a = aNext;
    b = bNext;
  }
}

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        def cut(node):
            slow = fast = node
            while fast and fast.next:
                slow, fast = slow.next, fast.next.next
            p = slow.next
            slow.next = None
            return p
          
          
        def reverse(node):
            prev = None
            while node:
                node.next, prev, node = prev, node, node.next
            return prev
          
          
        def merge(a, b):
            a_next = b_next = None
            while b:
                a_next = a.next
                b_next = b.next
                a.next = b
                b.next = a_next
                a = a_next
                b = b_next
                
                
        middle = cut(head)
        second = reverse(middle)
        merge(head, second)